app/src/main/java/protect/card_locker/importexport/CatimaImporter.java

```diff
@@ -362,22 +487,50 @@ private void importLoyaltyCard(Context context, SQLiteDatabase database, CSVReco
      * Import a single group into the database using the given
      * session.
      */
-    private void importGroup(SQLiteDatabase database, CSVRecord record) throws FormatException {
-        String id = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.ID, record, null);
+    private void importGroupV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.ID, record, null);
+        DBHelper.insertGroup(database, name);
+    }
 
-        DBHelper.insertGroup(database, id);
+    /**
+     * Import a single group from the V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private Pair<Integer, String> importGroupV3(SQLiteDatabase database, CSVRecord record) throws FormatException {
+        Integer id = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbGroups.ID, record, false);
+        String name = CSVHelpers.extractString(DBHelper.LoyaltyCardDbGroups.NAME, record, null);
+        DBHelper.insertGroup(database, name);
+        Pair<Integer, String> group = new Pair<>(id, name);
+        return group;
     }
 
     /**
      * Import a single card to group mapping into the database using the given
      * session.
      */
-    private void importCardGroupMapping(SQLiteDatabase database, CSVRecord record) throws FormatException {
+    private void importCardGroupMappingV2(SQLiteDatabase database, CSVRecord record) throws FormatException {
         Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
-        String groupId = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
+
+        String groupName = CSVHelpers.extractString(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, null);
+        cardGroups.add(DBHelper.getGroupByName(database, groupName));
+
+        DBHelper.setLoyaltyCardGroups(database, cardId, cardGroups);
+    }
 
+    /**
+     * Import a single card to group mapping from V3 scheme (database v16) into the database using the given
+     * session.
+     */
+    private void importCardGroupMappingV3(SQLiteDatabase database, CSVRecord record, Hashtable<Integer,String> groupsTable) throws FormatException {
+        Integer cardId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.cardID, record, false);
         List<Group> cardGroups = DBHelper.getLoyaltyCardGroups(database, cardId);
-        cardGroups.add(DBHelper.getGroup(database, groupId));
+
+        Integer groupId = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIdsGroups.groupID, record, false);
+        String groupName = groupsTable.get(groupId);
```

Comment:
I don't understand the whole `groupsTable` thing.

The export should contain the ID of the group, the card and the entry. You shouldn't have to do anything with the name at all during import.

Reply:
Should be solved now, removed the usage of groupName here as it was useless